src/main.c

```diff
@@ -1255,9 +1272,13 @@ emulator_loop(void *param)
 			continue;
 		}
 
-		uint32_t old_clockticks6502 = clockticks6502;
-		step6502();
-		uint8_t clocks = clockticks6502 - old_clockticks6502;
+		/* TODO: instead of using cpu_step, we should run CPU for more than one
+		   instruction (cpu_run) and rely on its cycle counter. this way
+		   the emulator loop can be made a lot tighter => faster.
+		   the CPU cycle counter can be read from the read/write callbacks.
+		   the only question is: how do we handle interrupts?
+		   see discussion in <https://github.com/commanderx16/x16-emulator/pull/437/> */
```

Comment:
Interrupts are only one of the problems here. While some components have "auto-step" implementations to catch up to the current CPU cycle on a memory read or write, not all components have this implementation (in particular those that are updated in this loop). You'd want to add appropriate implementations for those components. And some of these emulator-step components do so specifically because they're something like the VIA timers and they can generate an interrupt request at arbitrary intervals.

There are also potential problems with VERA replication. Granted, right now the VERA implementation has the idiosyncrasy that it renders an entire line when it does anything at all, but I suppose the point here is that the best-case scenario is to only run as many CPU cycles as it needed to reach the next scanline.

The best thought I've had on this, which I might eventually experiment with on Box16, is to have various components that can generate an IRQ implement some concept of estimating the number of clock ticks until the next IRQ is fired -- or perhaps estimate the clock tick counter value at which we expect the next IRQ to fire. In the latter case, some hypothetical `cpu_run()` would tick the CPU until it observed that the next expected IRQ clock tick counter value was `<=` the current clock tick counter value.

Reply:
(And in the case that some scheme like that were to be implemented, it would be totally acceptable for some component like VERA to lie in order to enforce opportunities to process its own internal logic -- so perhaps not an "estimated tick counter value for an IRQ" so much as a "please stop ticking the CPU here so I can run arbitrary logic (which in most cases is likely to generate an IRQ)."